[MathML] Ensure padded, fraction, underover and space elements update layout on attribute changes

Ahmad-S792 opened the pull request:
#### 424d7be45d8f9e6e9330b74cc0e174e57cde840a
<pre>
[MathML] Ensure padded, fraction, underover and space elements update layout on attribute changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=302575">https://bugs.webkit.org/show_bug.cgi?id=302575</a>
<a href="https://rdar.apple.com/164797996">rdar://164797996</a>

Reviewed by Frédéric Wang.

MathML &lt;mpadded&gt;, &lt;mfrac&gt;, &lt;munderover&gt;,  &lt;mover&gt; and &lt;mspace&gt; failed to
update width, height, depth, lspace, and voffset etc. when attributes
changed. The renderer continued to use stale cached metrics, causing
incorrect layout and unexpected spacing in incremental updates and dynamic
DOM modifications.

This patch resets the corresponding optional values and triggers
setNeedsLayoutAndPreferredWidthsUpdate() whenever affected attributes
change. This restores consistent layout recalculation across all
MathMLPaddedElement, MathMLFractionElement, MathMLSpaceElement and
MathMLUnderOverElement mutations.

Tests: mathml/non-core/frac-numalign-denomalign-002.html
       mathml/non-core/frac-numalign-denomalign-003.html

* Source/WebCore/mathml/MathMLFractionElement.cpp:
(WebCore::MathMLFractionElement::attributeChanged):
* Source/WebCore/mathml/MathMLPaddedElement.cpp:
(WebCore::MathMLPaddedElement::attributeChanged):
* Source/WebCore/mathml/MathMLSpaceElement.cpp:
(WebCore::MathMLSpaceElement::attributeChanged):
* Source/WebCore/mathml/MathMLUnderOverElement.cpp:
(WebCore::MathMLUnderOverElement::attributeChanged): Use `switch` here

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/fractions/frac-linethickness-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-001-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-002-expected.txt: Renamed from LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-002-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/scripts/underover-parameters-3-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/spaces/space-1-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-001-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-002-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/spaces/space-1-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-001-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/spaces/space-1-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-001-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-002-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/spaces/space-1-expected.txt: Removed.

Canonical link: <a href="https://commits.webkit.org/303508@main">https://commits.webkit.org/303508@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c8398355dd534201c2aba7c919103599ec03e9e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5093 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43662 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140119 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84613 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7387c257-171e-4f4f-808c-60eb82cd36a7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134468 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4852 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101368 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68664 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0e76972d-1544-4cf3-b9d4-2ff293d75c7c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3720 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118776 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82166 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3362a373-4769-4d22-8d45-f698a9a6eca8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3608 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1367 "Found 1 new API test failure: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83354 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112529 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36892 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142773 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4763 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37480 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109743 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4845 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4113 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109923 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27872 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3628 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115048 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58203 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4817 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33397 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4653 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68268 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4908 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4774 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebCore/mathml/MathMLPaddedElement.cpp:
```diff
@@ -77,26 +78,37 @@ const MathMLElement::Length& MathMLPaddedElement::voffset()
         break;
     }
     MathMLElement::attributeChanged(name, oldValue, newValue, attributeModificationReason);
+
```
nit: I would keep the call to the ancestor class' `attributeChanged` at the bottom of the function.

Ahmad-S792 replied:
Done.

fred-wang commented on Source/WebCore/mathml/MathMLSpaceElement.cpp:
```diff
@@ -66,20 +68,29 @@ const MathMLElement::Length& MathMLSpaceElement::depth()
     default:
         break;
     }
     MathMLPresentationElement::attributeChanged(name, oldValue, newValue, attributeModificationReason);
```
ditto.

Ahmad-S792 replied:
Done.

fred-wang reviewed (approved):
Great. Can you please confirm the 5 mpadded attributes and the 3 mspace attributes are covered by existing WPT tests in an independent way (i.e. for each `affectsLayout = true` line, removing it would at least make one of the WPT test fails)? If not, please extend WPT tests accordingly.

fred-wang commented on LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/fractions/fractions-dynamic-denomalign.html:
```diff
@@ -0,0 +1,79 @@
+<title>mfrac denomalign dynamic changes</title>
+<link rel="help" href="https://w3c.github.io/mathml-core/#fractions-mfrac">
+<link rel="help" href="https://w3c.github.io/mathml-core/#dom-and-javascript">
+<meta name="assert" content="Verifies dynamic denomalign application on mfrac.">
```
Note that numalign/denomalign are no longer parts of MathML Core and we have a test to check they have no effect: https://searchfox.org/wubkat/source/LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/fractions/frac-numalign-denomalign-001.html

It seems fine to have tests in WPT checking that the mfrac render the same after/before dynamic modification (if the attributes are ignored, that's still true!) but can you please explicitly use "dynamic modification of the deprecated denomalign attribute"? Also maybe this test should be renamed frac-numalign-denomalign-003.html

However, what you are doing is incorrect though you are checking left/center/right alignment after dynamic modification but in MathML Core they should always be centered. So maybe this should remain internal non-core tests, and maybe renamed LayoutTests/mathml/non-core/frac-numalign-denomalign-003.html ; also can you use `<!-- webkit-test-runner [ CoreMathMLEnabled=false ] -->`  so it works in platform that enabled MathML Core too?

Ahmad-S792 replied:
I will move it to internal and then add - `<!-- webkit-test-runner [ CoreMathMLEnabled=false ] -->`

Ahmad-S792 replied:
Addressed and moved the test to internal in `mathml/non-core`, renamed and also moved `Tests: ` bit in commit message above list of changes.

fred-wang commented on LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/fractions/fractions-dynamic-numalign.html:
```diff
@@ -0,0 +1,79 @@
+<title>mfrac numalign dynamic changes</title>
+<link rel="help" href="https://w3c.github.io/mathml-core/#fractions-mfrac">
+<link rel="help" href="https://w3c.github.io/mathml-core/#dom-and-javascript">
+<meta name="assert" content="Verifies dynamic numalign application on mfrac.">
```
Same comment as denomalign and I guess it should be called frac-numalign-denomalign-002.html

Ahmad-S792 replied:
Done.

fred-wang reviewed (approved):
Thanks!

Ahmad-S792 committed:
[MathML] Ensure padded, fraction, underover and space elements update layout on attribute changes
https://bugs.webkit.org/show_bug.cgi?id=302575
rdar://164797996

Reviewed by Frédéric Wang.

MathML <mpadded>, <mfrac>, <munderover>,  <mover> and <mspace> failed to
update width, height, depth, lspace, and voffset etc. when attributes
changed. The renderer continued to use stale cached metrics, causing
incorrect layout and unexpected spacing in incremental updates and dynamic
DOM modifications.

This patch resets the corresponding optional values and triggers
setNeedsLayoutAndPreferredWidthsUpdate() whenever affected attributes
change. This restores consistent layout recalculation across all
MathMLPaddedElement, MathMLFractionElement, MathMLSpaceElement and
MathMLUnderOverElement mutations.

Tests: mathml/non-core/frac-numalign-denomalign-002.html
       mathml/non-core/frac-numalign-denomalign-003.html

* Source/WebCore/mathml/MathMLFractionElement.cpp:
(WebCore::MathMLFractionElement::attributeChanged):
* Source/WebCore/mathml/MathMLPaddedElement.cpp:
(WebCore::MathMLPaddedElement::attributeChanged):
* Source/WebCore/mathml/MathMLSpaceElement.cpp:
(WebCore::MathMLSpaceElement::attributeChanged):
* Source/WebCore/mathml/MathMLUnderOverElement.cpp:
(WebCore::MathMLUnderOverElement::attributeChanged): Use `switch` here

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/fractions/frac-linethickness-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-001-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-002-expected.txt: Renamed from LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-002-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/scripts/underover-parameters-3-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/spaces/space-1-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-001-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-002-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/spaces/space-1-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-001-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/spaces/space-1-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-001-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/mpadded/mpadded-002-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/spaces/space-1-expected.txt: Removed.

Canonical link: https://commits.webkit.org/303508@main
